core/src/com/unciv/logic/automation/unit/UnitAutomation.kt

```diff
@@ -313,10 +316,17 @@ object UnitAutomation {
         if (unit.baseUnit.isRanged() && unit.hasUnique(UniqueType.HealsEvenAfterAction))
             return false // will heal anyway, and attacks don't hurt
 
-        if (tryPillageImprovement(unit)) return true
+        // Try pillage improvements until healed
+        while(tryPillageImprovement(unit, false)) {
+            if (unit.currentMovement == 0f || unit.health == 100) return true
```

Comment:
If I managed to heal to 100 and I still have movement, why should I return? I should keep going!

Reply:
Good point! That definitely makes sense.